crates/polars-parquet/src/arrow/write/primitive/basic.rs

```diff
@@ -26,20 +26,47 @@ where
     T: num_traits::AsPrimitive<P>,
 {
     if is_optional {
-        buffer.reserve(std::mem::size_of::<P>() * (array.len() - array.null_count()));
         // append the non-null values
-        for x in array.non_null_values_iter() {
-            let parquet_native: P = x.as_();
-            buffer.extend_from_slice(parquet_native.to_le_bytes().as_ref())
+        let validity = array.validity();
+
+        if let Some(validity) = validity {
+            let null_count = validity.unset_bits();
+
+            if null_count > 0 {
+                let values = array.values().as_slice();
+                let mut iter = validity.iter();
+
+                buffer.reserve(std::mem::size_of::<P>() * (array.len() - null_count));
+
+                let mut offset = 0;
+                let mut remaining_valid = array.len() - null_count;
+                while remaining_valid > 0 {
+                    let num_valid = iter.take_leading_ones();
```

Comment:
Nice trick with the `take_leading_ones`/`take_leading_zeros` alternation. Didn't know we had that. :)